src/GafferScene/PathFilter.cpp

```diff
@@ -166,17 +234,110 @@ void PathFilter::hashMatch( const ScenePlug *scene, const Gaffer::Context *conte
 	{
 		pathMatcherPlug()->hash( h );
 	}
+
+	if( rootsPlug()->getInput() )
+	{
+		rootSizesPlug()->hash( h );
+	}
 }
 
 unsigned PathFilter::computeMatch( const ScenePlug *scene, const Gaffer::Context *context ) const
 {
 	const ScenePlug::ScenePath &path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
 	ConstPathMatcherDataPtr pathMatcher = m_pathMatcher ? m_pathMatcher : pathMatcherPlug()->getValue();
-	return pathMatcher->readable().match( path );
+
+	if( !rootsPlug()->getInput() )
+	{
+		return pathMatcher->readable().match( path );
+	}
+
+	ConstIntVectorDataPtr rootSizes = rootSizesPlug()->getValue();
+	ScenePlug::ScenePath relativePath = path;
+	unsigned result = PathMatcher::NoMatch;
+
+	size_t previousRootSize = 0;
+	for( const int rootSize : rootSizes->readable() )
```

Comment:
If you wanted to be friendly to a new reader, it might be worth noting that rootSizes is inherently sorted with smallest first.

Reply:
https://github.com/GafferHQ/gaffer/pull/3692/commits/84790b5f7a61be35ab8ea1d2f3b532743cb73f36#diff-4716e8123851f6cde0e410f2ad84cbb2R84-R85